llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -13781,6 +13784,57 @@ struct NodeExtensionHelper {
   /// Check if this node needs to be fully folded or extended for all users.
   bool needToPromoteOtherUsers() const { return EnforceOneUse; }
 
+  void fillUpExtensionSupportForSplat(SDNode *Root, SelectionDAG &DAG,
+                                      const RISCVSubtarget &Subtarget) {
+    unsigned Opc = OrigOperand.getOpcode();
+    MVT VT = OrigOperand.getSimpleValueType();
+
+    assert((Opc == ISD::SPLAT_VECTOR || Opc == RISCVISD::VMV_V_X_VL) &&
+           "Unexpected Opcode");
+
+    if (Opc == ISD::SPLAT_VECTOR && !VT.isVector())
+      return;
+
+    // The pasthru must be undef for tail agnostic.
+    if (Opc == RISCVISD::VMV_V_X_VL && !OrigOperand.getOperand(0).isUndef())
+      return;
+
+    // Get the scalar value.
+    SDValue Op = Opc == ISD::SPLAT_VECTOR ? OrigOperand.getOperand(0)
+                                          : OrigOperand.getOperand(1);
+
+    // See if we have enough sign bits or zero bits in the scalar to use a
+    // widening opcode by splatting to smaller element size.
+    unsigned EltBits = VT.getScalarSizeInBits();
+    unsigned ScalarBits = Op.getValueSizeInBits();
+    // Make sure we're getting all element bits from the scalar register.
+    // FIXME: Support implicit sign extension of vmv.v.x?
+    if (ScalarBits < EltBits)
+      return;
+
+    unsigned NarrowSize = VT.getScalarSizeInBits() / 2;
+    // If the narrow type cannot be expressed with a legal VMV,
+    // this is not a valid candidate.
+    if (NarrowSize < 8)
+      return;
+
+    if (DAG.ComputeMaxSignificantBits(Op) <= NarrowSize)
+      SupportsSExt = true;
+
+    if (DAG.MaskedValueIsZero(Op,
+                              APInt::getBitsSetFrom(ScalarBits, NarrowSize)))
+      SupportsZExt = true;
+
+    EnforceOneUse = false;
+    CheckMask = Opc == ISD::SPLAT_VECTOR;
+
+    if (Opc == ISD::SPLAT_VECTOR)
+      std::tie(Mask, VL) =
+          getDefaultScalableVLOps(VT, SDLoc(Root), DAG, Subtarget);
```

Comment:
Not related to this PR, but neither splat_vector nor VXEXT_VL nor ISD::{SIGN,ZERO}_EXTEND have passthru operands. And we check that the passthru for VMV_V_X_VL is undef. So I'm not sure if we need to enforce that the mask and VL are the same here because the passthru will always be undef.

Reply:
> the mask and VL are the same

Can you please help explaining what is the meaning of this?

Reply:
Sorry that was poorly worded, I was referring to how we check that the root's mask and VL match each operand's mask and VL in `areVLAndMaskCompatible`. 

Although it looks like sometimes we take the mask/VL from the root, and sometime from the operand: https://github.com/llvm/llvm-project/blob/8b3b4a92adee40483c27f26c478a384cd69c6f05/llvm/lib/Target/RISCV/RISCVISelLowering.cpp#L13926-L13934

Is that correct?

Reply:
Opened #87997 which shows how it should be safe to remove the mask/vl checks